net/tstun/mtu.go

```diff
@@ -1,33 +1,168 @@
+// debugging. This value is used once at TUN creation and ignored
+// thereafter.
+//
+// User configured current MTU: The MTU set via the OS's tools
+// (ifconfig, ip, etc.). This MTU can change at any time. Setting the
+// MTU this way goes through the MTU() method of tailscale's TUN
+// wrapper.
+//
+// Maximum probed MTU: This is the largest MTU size that we send probe
+// packets for.
+//
+// Safe MTU: If the tailscale TUN MTU is set to this value, almost all
+// packets will get to their destination. Tailscale defaults to this
+// MTU in the absence of path MTU probe information or user MTU
+// configuration. We may occasionally find a path that needs a smaller
+// MTU but it is very rare.
+//
+// Peer MTU: This is the path MTU to a peer's current best
+// endpoint. It defaults to the Safe MTU unless we have path MTU probe
+// results that tell us otherwise.
+//
+// Initial MTU: This is the MTU tailscaled creates the TUN with. In
+// order of priority, it is:
+//
+// 1. If set, the value of TS_DEBUG_MTU clamped to a maximum of 65536
+// 2. If TS_DEBUG_ENABLE_PMTUD is set, the maximum size MTU we probe, minus wg overhead
+// 3. If TS_DEBUG_ENABLE_PMTUD is not set, the Safe MTU
+//
+// Current MTU: This the MTU of the tailscale TUN at any given moment
+// after TUN creation. In order of priority, it is:
+//
+// 1. The MTU set by the user via the OS, if it has ever been set
+// 2. If TS_DEBUG_ENABLE_PMTUD is set, the maximum size MTU we probe, minus wg overhead
+// 4. If TS_DEBUG_ENABLE_PMTUD is not set, the Safe MTU
+
+// TunMTU is used to describe MTUs for the tailscale TUN, while
+// WireMTU is used to describe MTUs for the underlying network
+// devices.
+type TunMTU uint32
+type WireMTU uint32
 
 const (
-	maxMTU     uint32 = 65536
-	defaultMTU uint32 = 1280
+	// Largest MTU we will consider for the Tailscale TUN.
+	maxTunMTU TunMTU = TunMTU(MaxPacketSize)
+	// A default "safe" MTU for the Tailscale TUN that we can use
+	// in the absence of other information such as path MTU probes.
+	safeTunMTU TunMTU = 1280
 )
 
-// DefaultMTU returns either the constant default MTU of 1280, or the value set
-// in TS_DEBUG_MTU clamped to a maximum of 65536.
-func DefaultMTU() uint32 {
-	// DefaultMTU is the Tailscale default MTU for now.
-	//
-	// wireguard-go defaults to 1420 bytes, which only works if the
-	// "outer" MTU is 1500 bytes. This breaks on DSL connections
-	// (typically 1492 MTU) and on GCE (1460 MTU?!).
-	//
-	// 1280 is the smallest MTU allowed for IPv6, which is a sensible
-	// "probably works everywhere" setting until we develop proper PMTU
-	// discovery.
-	tunMTU := defaultMTU
-	if mtu, ok := envknob.LookupUintSized("TS_DEBUG_MTU", 10, 32); ok {
-		mtu := uint32(mtu)
-		if mtu > maxMTU {
-			mtu = maxMTU
+// WireMTUsToProbe are the MTUs we are most likely to see in real world
+// networks. These are used to send discovery pings in packets padded
+// to exactly match the goal MTU.
+//
+// XXX Review these before committing
```

Comment:
PR is still a draft, but FYI that this is still here

Reply:
Thanks! Let me know if you have any thoughts on which sizes to test